ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/db/SlotCachingDasDb.java

```diff
@@ -0,0 +1,123 @@
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.spec.datastructures.blobs.versions.eip7594.DataColumnSidecar;
+import tech.pegasys.teku.spec.datastructures.networking.libp2p.rpc.DataColumnIdentifier;
+import tech.pegasys.teku.spec.datastructures.util.ColumnSlotAndIdentifier;
+
+/**
+ * The underlying DB primary key for a sidecar is {@link ColumnSlotAndIdentifier}. When a sidecar is
+ * requested by {@link DataColumnIdentifier} the slot for the block root needs to be queried This
+ * class serves two purposes:
+ *
+ * <ul>
+ *   <li>Optimizes extra call to the block DB when the slot is known from recent {@link
+ *       #addSidecar(DataColumnSidecar)} call
+ *   <li>Handles the case when a sidecar was stored optimistically without obtaining its
+ *       corresponding block. In this case a subsequent sidecar query by {@link
+ *       DataColumnIdentifier} would fail without this cache
+ * </ul>
+ */
+class SlotCachingDasDb extends AbstractDelegatingDasDb implements DataColumnSidecarDB {
+  private final DataColumnSidecarDB delegate;
+  private final ColumnSlotCache columnSlotCache = new ColumnSlotCache();
+
+  public SlotCachingDasDb(DataColumnSidecarDB delegate) {
+    super(delegate);
+    this.delegate = delegate;
+  }
+
+  @Override
+  public SafeFuture<Optional<DataColumnSidecar>> getSidecar(
+      DataColumnIdentifier dataColumnIdentifier) {
+    return columnSlotCache
+        .get(dataColumnIdentifier)
+        .map(delegate::getSidecar)
+        .orElseGet(() -> delegate.getSidecar(dataColumnIdentifier));
+  }
+
+  @Override
+  public SafeFuture<Void> addSidecar(DataColumnSidecar sidecar) {
+    columnSlotCache.addColumnSlotIdFromSidecar(sidecar);
+    return delegate.addSidecar(sidecar);
+  }
+
+  @Override
+  public SafeFuture<Void> setFirstCustodyIncompleteSlot(UInt64 slot) {
+    columnSlotCache.pruneCaches(slot);
+    return delegate.setFirstCustodyIncompleteSlot(slot);
+  }
+
+  // just delegate methods
+
+  @Deprecated
+  @Override
+  public SafeFuture<Optional<DataColumnSidecar>> getSidecar(ColumnSlotAndIdentifier identifier) {
+    return delegate.getSidecar(identifier);
+  }
+
+  @Override
+  public SafeFuture<Optional<UInt64>> getFirstCustodyIncompleteSlot() {
+    return delegate.getFirstCustodyIncompleteSlot();
+  }
+
+  @Override
+  public SafeFuture<Optional<UInt64>> getFirstSamplerIncompleteSlot() {
+    return delegate.getFirstSamplerIncompleteSlot();
+  }
+
+  @Override
+  public SafeFuture<Void> setFirstSamplerIncompleteSlot(UInt64 slot) {
+    return delegate.setFirstSamplerIncompleteSlot(slot);
+  }
+
+  @Override
+  public SafeFuture<Void> pruneAllSidecars(UInt64 tillSlot) {
+    return delegate.pruneAllSidecars(tillSlot);
+  }
+
+  private static class ColumnSlotCache {
+    private final Map<Bytes32, UInt64> blockRootToSlot = new HashMap<>();
+    private final NavigableMap<UInt64, Bytes32> slotToBlockRoot = new TreeMap<>();
```

Comment:
So it's not limited. Could it be dangerous on long non-finality streaks? Maybe limit it?

Reply:
Could it be enough to add `pruneCaches` on `pruneAllSidecars`? 
In this worst case the cache would be just around 4MB (assuming roughly 40 bytes per slot in cache)